libp2p/protocols/kademlia/kademlia.nim

```diff
@@ -162,35 +231,83 @@ proc refreshBuckets(kad: KadDHT) {.async: (raises: [CancelledError]).} =
       discard await kad.findNode(randomKey)
 
 proc maintainBuckets(kad: KadDHT) {.async: (raises: [CancelledError]).} =
-  heartbeat "refresh buckets", 10.minutes:
+  heartbeat "refresh buckets", chronos.minutes(10):
     await kad.refreshBuckets()
 
 proc new*(
-    T: typedesc[KadDHT], switch: Switch, rng: ref HmacDrbgContext = newRng()
+    T: typedesc[KadDHT],
+    switch: Switch,
+    validator: EntryValidator,
+    entrySelector: EntrySelector,
+    rng: ref HmacDrbgContext = newRng(),
 ): T {.raises: [].} =
   var rtable = RoutingTable.init(switch.peerInfo.peerId.toKey())
-  let kad = T(rng: rng, switch: switch, rtable: rtable)
+  let kad = T(
+    rng: rng,
+    switch: switch,
+    rtable: rtable,
+    dataTable: LocalTable.init(),
+    entryValidator: validator,
+    entrySelector: entrySelector,
+  )
 
   kad.codec = KadCodec
   kad.handler = proc(
       conn: Connection, proto: string
   ) {.async: (raises: [CancelledError]).} =
     try:
+      defer:
+        await conn.close()
       while not conn.atEof:
         let
           buf = await conn.readLp(MaxMsgSize)
           msg = Message.decode(buf).tryGet()
 
         case msg.msgType
         of MessageType.findNode:
-          let targetIdBytes = msg.key.get()
-          let targetId = PeerId.init(targetIdBytes).tryGet()
+          let targetIdBytes = msg.key.valueOr:
+            error "findNode message without key data present"
+            return
+          let targetId = PeerId.init(targetIdBytes).valueOr:
+            error "findNode message without valid key data"
+            return
           let closerPeers = kad.rtable.findClosest(targetId.toKey(), DefaultReplic)
           let responsePb = encodeFindNodeReply(closerPeers, switch)
           await conn.writeLp(responsePb.buffer)
 
           # Peer is useful. adding to rtable
           discard kad.rtable.insert(conn.peerId)
+        of MessageType.putValue:
+          var record =
+            if msg.record.isSome():
+              msg.record.unsafeGet()
+            else:
+              raise newException(CatchableError, "no record in message buffer")
+          let (skey, svalue) =
+            if record.key.isSome() and record.value.isSome():
+              (record.key.unsafeGet(), record.value.unsafeGet())
+            else:
+              raise newException(CatchableError, "no key or no value in rpc buffer")
+          let key = EntryKey(data: skey)
+          let value = EntryVal(data: svalue)
+          let ts = TimeStamp(ts: $times.now().utc)
+
+          let cand = EntryCandidate(key: key, value: value)
+          if not kad.entryValidator.validate(cand):
+            return
+          let others =
+            if kad.dataTable.entries.contains(key):
+              @[kad.dataTable.entries[key]]
+            else:
+              @[]
```

Comment:
There's a [`getOrDefault`](https://nim-lang.org/docs/tables.html#getOrDefault%2CTableRef%5BA%2CB%5D%2CA) method that I think could be used here instead of `if...else`

Reply:
problem there, is that get gives back a `Record` not a `seq[Record]` as needed. if you can suggest a way to ergonomacally do "get, and make it the only item in a sequence, otherwise the empty sequence", I'm all ears.